### PR TITLE
WIP Exp space 

### PR DESCRIPTION
**Pull request by under-Peter**
Now with diag/trace/ptrace

### CODE REVIEW

**Comment by under-Peter** [comment]
@GiggleLiu 
Now "expanding" einsums work, such as

`OMEinsum.einsum!(((),(2,)), (reshape([1]),[3, 4]), (2,2), zeros(2,2))`

work, i.e. "->jj". But I'll have to test it some more.
Next step should probably be adding the adjoints for AD, agree? Using Zygote?

**Comment by GiggleLiu** [comment]
Cool, It will finish the review as soon as possible.
Yes, since this interface will not change a lot, please go ahead.

File: src/einsum.jl [hunk 1]
```diff
@@ -79,3 +113,51 @@ function permuteandreshape(uniqueallins, t, c)
...
+    end
+    return stride
+end
+
+function densedelta(::Type{T}, ns::Vararg{Int,N}) where {T,N}
+    id = zeros(T,ns...)
+    o = one(T)
+    stride = deltastride(ns)
+    for i in 1:stride:length(id)
+        id[i] = o
```
Comment (GiggleLiu) [inline src/einsum.jl]: `@inbounds` (Julia 1.1+ compiler is quite clever, not sure this change will increase the performance)

File: src/einsum.jl [hunk 2]
```diff
@@ -79,3 +113,51 @@ function permuteandreshape(uniqueallins, t, c)
...
+    end
+    return id
+end
+
+function expandall!(b, indsb, a, indsa)
```
Comment (GiggleLiu) [inline src/einsum.jl]: Please add more comments/docstring to explain the motivation/algorithm.

File: test/runtests.jl [hunk 3]
```diff
@@ -1,80 +1,142 @@
...
-
-    # permutation
-    @test einsum(((1,2),), (a,), (2,1)) ≈ permutedims(a,(2,1))
-    @test einsum(((1,2,3,4),), (t,),(2,3,1,4)) ≈ permutedims(t,(2,3,1,4))
-
-    # tensor contraction
-    ta = zeros(size(t)[[1,2]]...)
-    for (i,j,k,l) in Iterators.product(1:2,1:2,1:2,1:2)
-        ta[i,l] += t[i,j,k,l] * a[j,k]
+    @testset "einsum" begin
```
Comment (GiggleLiu) [inline test/runtests.jl]: move these tests into another file like `einsum.jl` and incldue it in `runtests.jl`.

**Comment by GiggleLiu** [comment]
Since all issues have been resolved, the code is quite stable, with no apparent correctness and performance issues. I will merge it. Please checkout a new branch from master branch to add the AD functionality. 🎉

**Pull request merged**